feg/gateway/services/session_proxy/credit_control/virtual_apn_rules.go

```diff
@@ -15,43 +15,55 @@ package credit_control
 
 import (
 	"fmt"
-	"log"
 	"regexp"
 
 	"magma/feg/cloud/go/protos/mconfig"
+
+	"github.com/golang/glog"
 )
 
-func GetVirtualApnRule(apnFilter, apnOverwrite string) (*VirtualApnRule, error) {
-	reg, err := regexp.Compile(apnFilter)
+type VirtualApnRule struct {
+	ApnFilter                     *regexp.Regexp
+	ChargingCharacteristicsFilter *regexp.Regexp
+	ApnOverwrite                  string
+}
+
+func (v *VirtualApnRule) FromMconfig(pVirtualApnConfig *mconfig.VirtualApnRule) error {
+	apnRegex, err := regexp.Compile(pVirtualApnConfig.GetApnFilter())
+	if err != nil {
+		return fmt.Errorf("Failed to compile apn filter: %s", err)
+	}
+	ccRegex, err := regexp.Compile(pVirtualApnConfig.GetChargingCharacteristicsFilter())
 	if err != nil {
-		return nil, fmt.Errorf("Failed to compile apn filter: %s", err)
+		return fmt.Errorf("Failed to compile charging characteristics filter: %s", err)
 	}
-	return &VirtualApnRule{
-		ApnFilter:    reg,
-		ApnOverwrite: apnOverwrite,
-	}, nil
+	v.ApnFilter = apnRegex
+	v.ChargingCharacteristicsFilter = ccRegex
+	v.ApnOverwrite = pVirtualApnConfig.GetApnOverwrite()
+	return nil
 }
 
-func GenerateVirtualApnRules(config []*mconfig.VirtualApnRule) []*VirtualApnRule {
+func GenerateVirtualApnRules(pConfigs []*mconfig.VirtualApnRule) []*VirtualApnRule {
 	virtualApnConfigs := []*VirtualApnRule{}
-	for _, virtualApnlCfg := range config {
-		apnRule, err := GetVirtualApnRule(
-			virtualApnlCfg.GetApnFilter(),
-			virtualApnlCfg.GetApnOverwrite(),
-		)
+	for _, pVirtualApnCfg := range pConfigs {
+		apnRule := &VirtualApnRule{}
+		err := apnRule.FromMconfig(pVirtualApnCfg)
 		if err != nil {
-			log.Printf("%s Managed Gx Virtual APN Rule Config Load Error: %v", virtualApnlCfg.GetApnFilter(), err)
+			glog.Errorf("%s Managed Virtual APN Rule Config Load Error: %v", pVirtualApnCfg.GetApnFilter(), err)
 			continue
 		}
 		virtualApnConfigs = append(virtualApnConfigs, apnRule)
-		log.Printf("Virtual APN Rule Activated filter: %s, Overwrite: %s", virtualApnlCfg.GetApnFilter(), virtualApnlCfg.GetApnOverwrite())
+		glog.Infof("Virtual APN Rule Activated filter: %s, Overwrite: %s", pVirtualApnCfg.GetApnFilter(), pVirtualApnCfg.GetApnOverwrite())
```

Comment:
not sure if `GenerateVirtualApnRules` is triggered per each request or just at init. But if this is per request, then we should be setting this as `V(2)`

Reply:
This is done everytime the service starts. So Info is probably OK. :)